mypy/checkexpr.py

```diff
@@ -5124,6 +5124,15 @@ def apply_type_arguments_to_callable(
                     min_arg_count, len(type_vars), len(args), ctx
                 )
                 return AnyType(TypeOfAny.from_error)
+            for arg, tvar in zip(args, type_vars):
+                p = get_proper_type(arg)
+                if isinstance(p, AnyType):
+                    continue
+                if not isinstance(tvar, TypeVarType) or not tvar.values:
+                    continue
+                if not any(is_same_type(constraint, p) for constraint in tvar.values):
+                    # TODO: can we make the context the arg
+                    self.msg.incompatible_typevar_value(tp, arg, tvar.name, ctx, constrained=True)
```

Comment:
(should also have a better name for `p` but i left it because idk what it is)
```suggestion
            for arg, type_var in zip(args, type_vars):
                p = get_proper_type(arg)
                if isinstance(p, AnyType):
                    continue
                if not isinstance(tvar, TypeVarType) or not type_var.values:
                    continue
                if not any(is_same_type(constraint, p) for constraint in type_var.values):
                    # TODO: can we make the context the arg
                    self.msg.incompatible_typevar_value(tp, arg, type_var.name, ctx, constrained=True)
```